src/events/requirer.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for related applications on the `zookeeper` relation interface."""
+import logging
+from typing import TYPE_CHECKING
+
+from ops.charm import RelationBrokenEvent, RelationEvent
+from ops.framework import Object
+
+from literals import OPENSEARCH_REL_NAME
+
+if TYPE_CHECKING:
+    from charm import OpensearchDasboardsCharm
+
+logger = logging.getLogger(__name__)
+
+
+class RequirerEvents(Object):
+    """Event handlers for related applications on the `zookeeper` relation interface."""
+
+    def __init__(self, charm):
+        super().__init__(charm, "provider")
+        self.charm: "OpensearchDasboardsCharm" = charm
+
+        self.framework.observe(
+            self.charm.on[OPENSEARCH_REL_NAME].relation_changed, self._on_client_relation_updated
+        )
+        self.framework.observe(
+            self.charm.on[OPENSEARCH_REL_NAME].relation_broken, self._on_client_relation_broken
+        )
+
+    def _on_client_relation_updated(self, event: RelationEvent) -> None:
+        """Updates ACLs while handling `client_relation_updated` events."""
+        if not self.charm.state.stable:
+            event.defer()
+            return
+
+        if (
+            (
+                self.charm.client_requires_interface.fetch_relation_field(
```

Comment:
I'm sure that the current construct will benefit from lazy evaluation, skipping conditions if earlier ones already determine the result.

I'm not sure that the same would happen when using variables (across Python versions?) thus I'm thinking it may be safest the way it is.

Feel free to challange the above in case I'm mistaken/ignorant on this.

Reply:
I understand your view and it makes sense, I do however believe that the readability gains outweigh the performance gains in this case

Reply:
I honestly don't think readability would suffer a severe harm here. Code looks "ugly", yet I find it readable. However this is a subjective matter :-)

And thought I'd disagree that it reaches the priority compared to performance -- I also realize that my performance concerns were so minimal, that we evaluate properties in question a million times through an event handler :sweat_smile: 

No matter what the debate is solved in a different way. Namely that I realized that I don't need 2 manually fetch this data, as `self.charm.opensearch_server.<property>` interface attributes are refeering precisely to this. Thus the code simplifies and also keeps "performance" considerations :-)